components/concurrency_manager/src/lib.rs

```diff
@@ -52,9 +101,70 @@ impl ConcurrencyManager {
 
     /// Updates max_ts with the given new_ts. It has no effect if
     /// max_ts >= new_ts or new_ts is TimeStamp::max().
-    pub fn update_max_ts(&self, new_ts: TimeStamp) {
-        if new_ts != TimeStamp::max() {
-            self.max_ts.fetch_max(new_ts.into_inner(), Ordering::SeqCst);
+    ///
+    /// To avoid invalid ts breaking the invariants, the new_ts should be
+    /// less than or equal to the max_ts_limit.
+    ///
+    /// # Returns
+    /// - Ok(()): If the update is successful or has no effect
+    /// - Err(limit): If new_ts is greater than the max_ts_limit, returns the
+    ///   current limit value
+    pub fn update_max_ts(
+        &self,
+        new_ts: TimeStamp,
+        source: Option<String>,
+    ) -> Result<(), InvalidMaxTsUpdate> {
+        if new_ts == TimeStamp::max() {
+            return Ok(());
+        }
+        let new_ts = new_ts.into_inner();
+        let limit = self.max_ts_limit.load(Ordering::SeqCst);
+        if limit > 0 && new_ts > limit {
+            let since_last_limit_update_ms = self.start_instant.saturating_elapsed().as_millis()
+                as u64
+                - self.last_update_limit_instant.load(Ordering::SeqCst);
+            if since_last_limit_update_ms < self.limit_valid_duration.as_millis() as u64 {
+                let source = source.unwrap_or_default();
+                error!("invalid max_ts update";
+                    "attempted_ts" => new_ts,
+                    "max_allowed" => limit,
+                    "source" => &source
+                );
+                if self.panic_on_invalid_max_ts.load(Ordering::SeqCst) {
+                    panic!(
+                        "invalid max_ts update: {} exceeds the limit {}, source={}",
+                        new_ts, limit, source
+                    );
+                }
+                return Err(InvalidMaxTsUpdate {
+                    attempted_ts: TimeStamp::new(new_ts),
+                    max_allowed: TimeStamp::new(limit),
+                });
+            }
+        }
+        MAX_TS_GAUGE.set(self.max_ts.fetch_max(new_ts, Ordering::SeqCst).max(new_ts) as i64);
+        Ok(())
+    }
```

Comment:
@ekexium @MyonKeminta 
For the else branch, how about comparing the `new_ts` with `local_now + allowed_gap` and print error logs if the `new_ts` > `local_now + allowed_gap`.

Reply:
Yes, and just print error logs when the approximate check fails and do not panic.

Reply:
Do you think it should report an error to user when using approximate?

Reply:
Reporting error is fine to me as it is an unexpected abnormal phenomenon considering the clock drift.

BTW, we still need to consider the risk that a misjudgment could lead to the risk of certain calls continuously reporting errors and being unavailable.